spqlios/arithmetic/vec_znx_convolution.c

```diff
@@ -0,0 +1,150 @@
+) {
+  return module->func.cnv_prepare_left_contiguous(module, pvec, nrows, a, a_size, a_sl, tmp_space);
+}
+
+/** @brief minimal scratch space byte-size required for the cnv_prepare_left_contiguous function */
+EXPORT uint64_t cnv_prepare_left_contiguous_tmp_bytes(const MODULE* module,  // N
+                                                      uint64_t nrows,        // size of output
+                                                      uint64_t a_size        // size of input
+) {
+  return module->func.cnv_prepare_left_contiguous_tmp_bytes(module, nrows, a_size);
+}
+
+/** @brief prepares the right vector for convolution  */
+EXPORT void fft64_convolution_prepare_right_contiguous_ref(const MODULE* module,                              // N
+                                                           CNV_PVEC_R* pvec, uint64_t nrows,                  // output
+                                                           const int64_t* a, uint64_t a_size, uint64_t a_sl,  // a
+                                                           uint8_t* tmp_space  // scratch space
+) {
+  fft64_convolution_prepare_contiguous_ref(module, (double*)pvec, nrows, a, a_size, a_sl, tmp_space);
+}
+
+/** @brief minimal scratch space byte-size required for the cnv_prepare_left_contiguous function */
+EXPORT uint64_t fft64_convolution_prepare_left_contiguous_tmp_bytes(const MODULE* module,  // N
+                                                                    uint64_t nrows,        // size of output
+                                                                    uint64_t a_size        // size of input
+) {
+  const uint64_t nn = module->nn;
+  const uint64_t rows = nrows < a_size ? nrows : a_size;
+  return nn * sizeof(int64_t) * rows;
+}
+
+/** @brief minimal scratch space byte-size required for the cnv_prepare_right_contiguous function */
+EXPORT uint64_t fft64_convolution_prepare_right_contiguous_tmp_bytes(const MODULE* module,  // N
+                                                                     uint64_t nrows,        // size of output
+                                                                     uint64_t a_size        // size of input
+) {
+  uint64_t nn = module->nn;
+  const uint64_t rows = nrows < a_size ? nrows : a_size;
+  return nn * sizeof(uint64_t) * rows;
+}
+
+/** @brief prepares the left vector for convolution  */
+EXPORT void fft64_convolution_prepare_left_contiguous_ref(const MODULE* module,                              // N
+                                                          CNV_PVEC_L* pvec, uint64_t nrows,                  // output
+                                                          const int64_t* a, uint64_t a_size, uint64_t a_sl,  // a
+                                                          uint8_t* tmp_space  // scratch space
+) {
+  fft64_convolution_prepare_contiguous_ref(module, (double*)pvec, nrows, a, a_size, a_sl, tmp_space);
```

Comment:
ideally, `cnv_prepare_left` should divide the result by N compared to `cnv_prepare_right`

indeed, it will compensate the fact that the fft product multiplies everything by N.